webui/react/src/components/DropdownMenu.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { PropsWithChildren, useCallback, useEffect, useMemo, useState } from 'react';
+
+import { toRem } from 'utils/dom';
+import { generateLetters } from 'utils/string';
+
+import css from './DropdownMenu.module.scss';
+
+export enum Placement {
+  Bottom = 'bottom',
+  BottomLeft = 'bottomLeft',
+  BottomRight = 'bottomRight',
+  Left = 'left',
+  LeftBottom = 'leftBottom',
+  LeftTop = 'leftTop',
+  Right = 'right',
+  RightTop = 'rightTop',
+  RightBottom = 'rightBottom',
+  Top = 'top',
+  TopLeft = 'topLeft',
+  TopRight = 'topRight',
+}
+
+interface Props {
+  menu: React.ReactNode;
+  offset?: { x: number, y: number };
+  onVisibleChange?: (visible: boolean) => void;
+  placement?: Placement;
+  showArrow?: boolean;
+}
+
+const DropdownMenu: React.FC<Props> = ({
+  offset = { x: 0, y: 0 },
+  onVisibleChange,
+  placement = Placement.BottomLeft,
+  showArrow = true,
+  ...props
+}: PropsWithChildren<Props>) => {
+  const [ id ] = useState(generateLetters());
+  const [ isVisible, setIsVisible ] = useState(false);
+  const classes = [ css.base, css[placement] ];
+
+  const menuStyle = useMemo(() => {
+    switch (placement) {
+      case Placement.Bottom:
+        return {
```

Comment:
it'd be nice if we didn't have to repeat these but I don't see a good abstraction. maybe just `calc(100% + ${toRem(offset.y)})`

Reply:
I didn't have a better way to abstract these. Even the `calc(100% + ${toRem(offset.y)})` was only applicable to the bottom placements. I had thought about adding properties based on Left, Middle, Right and Top, Middle, Bottom, but that doesn't really save much any lines and it just becomes much harder to read and figure out.